extensions/cms/src/cmsUtils.ts

```diff
@@ -79,40 +105,65 @@ export class CmsUtils {
 		});
 	}
 
+	public static didConnectionChange(connectionA: azdata.connection.Connection, connectionB: azdata.connection.Connection): boolean {
+		return (connectionA !== connectionB) || ((connectionA.connectionId === connectionB.connectionId) &&
+			(connectionA.options.savePassword !== connectionA.options.savePassword));
+	}
+
 	public async createCmsServer(connection: azdata.connection.Connection,
-		name: string, description: string): Promise<mssql.ListRegisteredServersResult> {
+		name: string, description: string): Promise<CreateCmsResult> {
 		let provider = await this.getCmsService();
 		connection.providerName = connection.providerName === cmsProvider ? mssqlProvider : connection.providerName;
 		let ownerUri = await azdata.connection.getUriForConnection(connection.connectionId);
 		if (!ownerUri) {
 			// Make a connection if it's not already connected
-			await azdata.connection.connect(Utils.toConnectionProfile(connection), false, false).then(async (result) => {
+			let initialConnectionProfile = CmsUtils.getConnectionProfile(connection);
+			await azdata.connection.connect(initialConnectionProfile, false, false).then(async (result) => {
 				ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
+				// If the ownerUri is still undefined, then open a connection dialog with the connection
+				if (!ownerUri) {
+					await this.connection(initialConnectionProfile).then(async (result) => {
+						if (result) {
+							ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
+							connection = result;
+						}
+					}, (error) => {
+						// cancel pressed on connection dialog
+						return Promise.reject(localize('cms.error.cancelConnectionDialog', 'The server is disconnected. Please connect to the server to continue.'));
+					});
```

Comment:
This error message doesn't seem to match what you're saying is happening here

Reply:
This happens when you try to expand the server and if the password isn't saved. The connection dialog pops up, but if you cancel, we show this. It makes sense to me.

Reply:
Why does cancelling the dialog mean the server is disconnected?

Reply:
Because it's not connected until you connect it from the connection dialog. It's exactly like the SSMS behavior if you don't know what's going on.

Reply:
Is this error message exposed to the user? If it is I would disagree that this is a reasonable message to give. If I click cancel why would it tell me something about the server being disconnected? Why would we even be creating the server in the first place? It might help if you went over this with me in person if you think I'm not understanding the problem correctly

---
In reply to: [292675076](https://github.com/microsoft/azuredatastudio/pull/5989#discussion_r292675076) [](ancestors = 292675076)